client/src/main/java/io/pravega/client/stream/impl/EventStreamReaderImpl.java

```diff
@@ -154,8 +156,8 @@
         } 
         lastRead = Sequence.create(segment.getSegmentId(), offset);
         int length = buffer.remaining() + WireCommands.TYPE_PLUS_LENGTH_SIZE;
-        return new EventReadImpl<>(deserializer.deserialize(buffer), getPosition(),
-                                   new EventPointerImpl(segment, offset, length), null);
+        lastPosition = getPosition();
```

Comment:
Using a member variable is error prone. (In fact there is a case this logic does not cover: if the segments change as a result of calling updateGroupStateIfNeeded, the next time through the above while loop it will pass in stale data resulting in an invalid state.)

Instead it may be better to just use a local variables by computing the position on 128 an passing it into updateGroupStateIfNeeded, then modifying the offset for the segment read from before returning it.

Reply:
Maybe I'm mistaken, but your suggestion implies that we should be able to "update" the offsets of a `PositionImpl` object in some way, so we don't have to instantiate it twice anyway, right? I don't know if this is possible at the moment, as this object seems to be read-only and uses unmodifiable collections in its getters. 

But your comment made me think on a more aggressive optimization: let's assume that we still keep a member variable for the "last position" object. As you mention, we fully recompute the last position object whenever there is a change in the segments, so the state is valid upon segment changes. If there are no segment changes, which is the majority of cases, we can modify the offset of the segment read on a particular iteration (which would imply to be able to be able to safely update the Position object) reusing the "last position" object. If we do that, we would be able to avoid 99% of the `getPosition()` execution calls, which is the main performance bottleneck we have now in the reader side. Will try to propose this change so we can discuss about it.

Reply:
I think we can avoid this. See my comments below.